index.d.ts

```diff
@@ -0,0 +1,60 @@
+export interface Action {
+  type: string;
+}
+
+export type Reducer<S> = (state: S, action: Action) => S;
+
+export type Dispatch = (action: any) => any;
```

Comment:
I see how your proposal can be useful: if we put
```ts
type Dispatch<A, O> = (action: A) => O;
```
Then original `Store` definition (no middleware) would be:
```ts
interface Store<S> {
  dispatch: Dispatch<Action, Action>;
}
```
Then we can strengthen `Middleware` type something like:
```ts
interface Middleware {
  <S, A, O>(api: MiddlewareAPI<S>): 
    <D extends Dispatch<any, any>>(next: D) => D & Dispatch<A, O>;
}
``` 
Thus we could always have strong typings for `dispatch` after all store enhancers were applied.
I need to experiment with this a bit.